modules/paginator/paginator.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+// Copyright 2015 Unknwon. Licensed under the Apache License, Version 2.0
+
+package paginator
+
+/*
+In template:
+
+```html
+{{if not .Page.IsFirst}}[First](1){{end}}
+{{if .Page.HasPrevious}}[Previous]({{.Page.Previous}}){{end}}
+
+{{range .Page.Pages}}
+	{{if eq .Num -1}}
+	...
+	{{else}}
+	{{.Num}}{{if .IsCurrent}}(current){{end}}
+	{{end}}
+{{end}}
+
+{{if .Page.HasNext}}[Next]({{.Page.Next}}){{end}}
+{{if not .Page.IsLast}}[Last]({{.Page.TotalPages}}){{end}}
+```
+
+Output:
+
+```
+[First](1) [Previous](2) ... 2 3(current) 4 ... [Next](4) [Last](5)
+```
+*/
+
+// Paginator represents a set of results of pagination calculations.
+type Paginator struct {
+	total     int // total rows count
+	pagingNum int // how many rows in one page
+	current   int // current page number
+	numPages  int // how many pages to show on the UI
+}
+
+// New initialize a new pagination calculation and returns a Paginator as result.
+func New(total, pagingNum, current, numPages int) *Paginator {
+	if pagingNum <= 0 {
+		pagingNum = 1
+	}
+	if current <= 0 {
+		current = 1
+	}
+	p := &Paginator{total, pagingNum, current, numPages}
+	if p.current > p.TotalPages() {
+		p.current = p.TotalPages()
+	}
+	return p
+}
+
+// IsFirst returns true if current page is the first page.
+func (p *Paginator) IsFirst() bool {
+	return p.current == 1
+}
+
+// HasPrevious returns true if there is a previous page relative to current page.
+func (p *Paginator) HasPrevious() bool {
+	return p.current > 1
+}
+
+func (p *Paginator) Previous() int {
+	if !p.HasPrevious() {
+		return p.current
+	}
+	return p.current - 1
+}
+
+// HasNext returns true if there is a next page relative to current page.
+func (p *Paginator) HasNext() bool {
+	return p.total > p.current*p.pagingNum
+}
+
+func (p *Paginator) Next() int {
+	if !p.HasNext() {
+		return p.current
+	}
+	return p.current + 1
+}
+
+// IsLast returns true if current page is the last page.
+func (p *Paginator) IsLast() bool {
+	if p.total == 0 {
+		return true
+	}
+	return p.total > (p.current-1)*p.pagingNum && !p.HasNext()
+}
+
+// Total returns number of total rows.
+func (p *Paginator) Total() int {
+	return p.total
+}
+
+// TotalPages returns number of total pages.
+func (p *Paginator) TotalPages() int {
+	if p.total == 0 {
+		return 1
+	}
+	return (p.total + p.pagingNum - 1) / p.pagingNum
```

Comment:
```suggestion
	return (p.total - 1)/p.pagingNum + 1
```
Doing the + pagingNum is effectively just adding a one to the end result. So bit useless to do in that way(I know it's copied from the package, but since were having it's own module now...)

Reply:
It's not copied from the package, it was `if ...`, I wrote the expression 😂

Why I always use `(total + align - 1) /  align` because the `total` can be unsigned in some cases, `total - 1` then would overflow. And the `total` can never reach around UNSIGNED_INTxx_MAX (it's impossible to hold so much data in real world).

Especially when using C++, all size_t are unsigned, so the only choice is using `(total + align - 1) /  align` there.

In conclusion, I think  `(total + align - 1) /  align` is a very general and safe method to do the math trick, and I hope we can keep it.

Reply:
> Why I always use `(total + align - 1) / align` because the `total` can be unsigned in some cases, `total - 1` then would overflow. And the `total` can never reach around UNSIGNED_INTxx_MAX (it's impossible to hold so much data in real world).
> 
> Especially when using C++, all size_t are unsigned, so the only choice is using `(total + align - 1) / align` there.
> 
> In conclusion, I think `(total + align - 1) / align` is a very general and safe method to do the math trick, and I hope we can keep it.

But you could overflow if align is some insane high number(close to math.MaxInt64) the suggested change simply avoids that 😄. But it's a edge-case which we don't really have to worry about.